File: web/ack.js
// Receiver-side acknowledgement generation (mirrors server/ack.py).
// mode "packet": one ACK per DATA packet. mode "block": ACK_BLOCK when every_n
// packets are pending or interval_ms after the first pending packet.

import * as proto from './protocol.js';

const RANGE_WINDOW = 4096;

export class AckGenerator {
  constructor(flow, config, send, nowMs) {
    config = config ?? {};
    this.flow = flow;
    this.mode = config.mode ?? 'packet';
    this.intervalMs = config.interval_ms ?? 20;
    this.everyN = config.every_n ?? 16;
    this.send = send;
    this.nowMs = nowMs;
    this.received = new Set();
    this.largest = -1;
    this.largestRecvTs = 0;
    this.pending = [];
    this.ackId = 0;
    this.acksSent = 0;
    this.timer = null;
  }

  onPacket(seq, sendTs, recvTs) {
    if (this.mode === 'packet') {
      this.send(proto.encodeAck(this.flow, seq, sendTs, recvTs));
      this.acksSent++;
      return;
    }
    this.received.add(seq);
    if (seq > this.largest) { this.largest = seq; this.largestRecvTs = recvTs; }
    this.pending.push([seq, recvTs]);
    if (this.pending.length >= this.everyN) this.flush();
    else if (this.timer === null) this.timer = setTimeout(() => this.flush(), this.intervalMs);
  }

  ranges() {
    const floor = Math.max(0, this.largest - RANGE_WINDOW + 1);
    const ranges = [];
    let seq = this.largest;
    while (seq >= floor && ranges.length < proto.MAX_ACK_RANGES) {
      const last = seq;
      while (seq >= floor && this.received.has(seq)) seq--;
      ranges.push([seq + 1, last]);
      while (seq >= floor && !this.received.has(seq)) seq--;
    }
    const low = seq >= floor ? ranges[ranges.length - 1][0] : floor;
    if (this.received.size > 2 * RANGE_WINDOW) {
      for (const s of this.received) if (s < floor) this.received.delete(s);
    }
    return { ranges, low };
  }

  flush() {
    if (this.timer !== null) { clearTimeout(this.timer); this.timer = null; }
    if (this.mode !== 'block' || this.largest < 0) return;
    const { ranges, low } = this.ranges();
    const capacity = proto.blockTsCapacity(ranges.length);
    const pending = this.pending;
    this.pending = [];
    const delay = Math.max(0, this.nowMs() - this.largestRecvTs);
    const chunks = [];
    for (let i = 0; i < pending.length; i += capacity) chunks.push(pending.slice(i, i + capacity));
    if (!chunks.length) chunks.push([]);
    for (const chunk of chunks) {
      this.send(proto.encodeAckBlock(this.flow, this.ackId++, this.largest, this.largestRecvTs, delay, low, ranges, chunk));
      this.acksSent++;
    }
  }

  close() {
    if (this.timer !== null) { clearTimeout(this.timer); this.timer = null; }
  }
}

File: server/ack.py
"""Receiver-side acknowledgement generation (mirrored in web/ack.js).

mode "packet": one ACK datagram per received DATA packet (echoes send_ts).
mode "block":  ACK_BLOCK datagrams, sent when `every_n` packets are pending or
               `interval_ms` after the first pending packet, whichever is first.
mode "none":   no ACKs at all. For saturation tests, where the sender is trying
               to fill its uplink and ACK traffic would only add work at both
               ends; the server sees every packet anyway.
"""

import asyncio
from typing import Callable, Dict, List, Optional, Set, Tuple

import protocol as proto

RANGE_WINDOW = 4096  # sequence numbers below largest considered for ranges


class AckGenerator:
    def __init__(
        self,
        flow: int,
        config: Optional[Dict],
        send: Callable[[bytes], None],
        now_ms: Callable[[], float],
    ) -> None:
        config = config or {}
        self.flow = flow
        self.mode = config.get("mode", "packet")
        self.interval_ms = float(config.get("interval_ms", 20))
        self.every_n = int(config.get("every_n", 16))
        self.send = send
        self.now_ms = now_ms

        self.received: Set[int] = set()
        self.largest = -1
        self.largest_recv_ts = 0.0
        self.pending: List[Tuple[int, float]] = []
        self.ack_id = 0
        self.acks_sent = 0
        self._timer: Optional[asyncio.TimerHandle] = None

    def on_packet(self, seq: int, send_ts: float, recv_ts: float) -> None:
        if self.mode == "none":
            return
        if self.mode == "packet":
            self.send(proto.encode_ack(self.flow, seq, send_ts, recv_ts))
            self.acks_sent += 1
            return

        self.received.add(seq)
        if seq > self.largest:
            self.largest, self.largest_recv_ts = seq, recv_ts
        self.pending.append((seq, recv_ts))
        if len(self.pending) >= self.every_n:
            self.flush()
        elif self._timer is None:
            self._timer = asyncio.get_event_loop().call_later(self.interval_ms / 1000, self.flush)

    def _ranges(self) -> Tuple[List[Tuple[int, int]], int]:
        floor = max(0, self.largest - RANGE_WINDOW + 1)
        ranges: List[Tuple[int, int]] = []
        seq = self.largest
        while seq >= floor and len(ranges) < proto.MAX_ACK_RANGES:
            last = seq
            while seq >= floor and seq in self.received:
                seq -= 1
            ranges.append((seq + 1, last))
            while seq >= floor and seq not in self.received:
                seq -= 1
        # Stopped early: only [first of last range, largest] is described.
        low = ranges[-1][0] if seq >= floor else floor
        # Forget sequence numbers that can no longer appear in a range.
        if len(self.received) > 2 * RANGE_WINDOW:
            self.received = {s for s in self.received if s >= floor}
        return ranges, low

    def flush(self) -> None:
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None
        if self.mode != "block" or self.largest < 0:
            return
        ranges, low = self._ranges()
        capacity = proto.block_ts_capacity(len(ranges))
        pending, self.pending = self.pending, []
        delay = max(0.0, self.now_ms() - self.largest_recv_ts)
        chunks = [pending[i : i + capacity] for i in range(0, len(pending), capacity)] or [[]]
        for chunk in chunks:
            self.send(
                proto.encode_ack_block(
                    self.flow, self.ack_id, self.largest, self.largest_recv_ts, delay, low, ranges, chunk
                )
            )
            self.ack_id += 1
            self.acks_sent += 1

    def close(self) -> None:
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None
